### PR TITLE
Translate LLVM-IR zero-length arrays to 1-length arrays in SPIR-V

### PR DESCRIPTION
**Pull request by lbushi25**
The LLVM-IR type of zero-length array does not seem to have a mapping into SPIR-V type, rather it outputs an error since the obvious choice of zero-length array is not valid in SPIR-V. This, for example, prohibits us from using unbounded arrays in SYCL device kernels because they are typically represented in LLVM-IR through zero-length arrays. This PR attempts to introduce a workaround to this by lowering a zero-length array in LLVM-IR to a 1-length array in SPIR-V.

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp
Comment (lbushi25) [inline lib/SPIRV/SPIRVWriter.cpp:1]: Only line 452-469 have functional changes in this file. The rest are formatter artifacts.

**Review by svenvh — CHANGES_REQUESTED** [review]
Unconditionally generating non-conformant SPIR-V seems not right.  The suggestion in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/417 was to hide such functionality behind an option (or perhaps we should consider a SPIR-V extension?).

**Comment by lbushi25** [comment]
> Unconditionally generating non-conformant SPIR-V seems not right. The suggestion in #417 was to hide such functionality behind an option (or perhaps we should consider a SPIR-V extension?).

Why would this be non-conformant SPIR-V?
I'm not generating arrays of size zero. I'm instead modifying the lowering of LLVM IR arrays of zero-length to SPIR-V arrays of length 1 which are, at least to the best of my knowledge, valid in SPIR-V

**Comment by svenvh** [comment]
> Why would this be non-conformant SPIR-V? I'm not generating arrays of size zero. I'm instead modifying the lowering of LLVM IR arrays of zero-length to SPIR-V arrays of length 1 which are, at least to the best of my knowledge, valid in SPIR-V.

My apologies, I completely misread your initial description.  Lowering to a SPIR-V array of length 1 is of course completely conformant.

File: lib/SPIRV/SPIRVWriter.cpp
Comment (svenvh) [reply lib/SPIRV/SPIRVWriter.cpp:1]: Ideally we should leave out the formatter artifacts as they may complicate potential backports of this commit.  This repo ought to be clang-formatted, so we should fix the formatting in a separate PR; I can do so if you wish.

Comment (lbushi25) [reply lib/SPIRV/SPIRVWriter.cpp:1]: > Ideally we should leave out the formatter artifacts as they may complicate potential backports of this commit. This repo ought to be clang-formatted, so we should fix the formatting in a separate PR; I can do so if you wish.
> 
> Update: I've created #2744 . When that has landed, you can just rebase on top of `main` and the unrelated formatting changes should disappear from your patch.

Thanks for creating the formatting patch. I have merged it into my branch and also made the changes you suggested.

**Review by svenvh — APPROVED** [review]

**Review by MrSidims — APPROVED** [review]

**Pull request merged**